### PR TITLE
Multi-modal RAG template

### PR DESCRIPTION
**Pull request by rlancemartin**
* OpenCLIP embeddings
* GPT-4V

### CODE REVIEW

File: templates/rag-chroma-multi-modal/rag_chroma_multi_modal/chain.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+image_uris = sorted(
+    [
+        os.path.join(rel_img_dump_path, image_name)
+        for image_name in os.listdir(rel_img_dump_path)
+        if image_name.endswith(".jpg")
+    ]
+)
+
+# Add images
+vectorstore_mmembd.add_images(uris=image_uris)
```
Comment (efriis) [inline templates/rag-chroma-multi-modal/rag_chroma_multi_modal/chain.py:145]: let's get in the habit of doing this in an `ingest.py` script instead, so it doesn't run every time the langserve instance restarts

Comment (efriis) [reply templates/rag-chroma-multi-modal/rag_chroma_multi_modal/chain.py:145]: (and then reference that you have to run `python ingest.py` in the readme)

**Review by efriis — COMMENTED** [review]
Running into issues with gpt-4v I think: https://smith.langchain.com/public/007ba159-3549-4286-b4c0-91c4ad7c4310/r

**Comment by rlancemartin** [comment]
> https://smith.langchain.com/public/007ba159-3549-4286-b4c0-91c4ad7c4310/r

Right, prompt tokens are > 100k. 

1/ It looks like it retrieves 4 images [here](https://smith.langchain.com/public/007ba159-3549-4286-b4c0-91c4ad7c4310/r/9d45b5cd-2693-4521-b407-ac74933863e0).

2/ But, as expected, we restrict to at max [2 images](https://smith.langchain.com/public/007ba159-3549-4286-b4c0-91c4ad7c4310/r/abaa6e09-27f8-4cc4-ae3e-6c456500b312) in the prompt.

3/ Those 2 are passed to GPT-4V [here](https://smith.langchain.com/public/007ba159-3549-4286-b4c0-91c4ad7c4310/r/a9db9a3f-e1e7-4af2-b779-827e27d30fc4), which reports `111,332 tokens`.

4/ Previously, I passed the same 2 images [here](https://smith.langchain.com/public/7a0a5813-b3c0-4e37-a362-eeb6fa4e3bdc/r/b29322fe-bc0b-4ed3-a624-66cd3f87a31d), and GPT-4V reports `2,504 tokens`.

The image size should be consistent; [before](https://github.com/langchain-ai/langchain-benchmarks/blob/4198cdff18f2609526a8f63fc91bbcfeaeb18b63/docs/source/notebooks/retrieval/multi_modal_benchmarking/multi_modal_eval.ipynb) I also used `size=(1280, 720)`.

This is a weird bug that I have noticed before, but have not been able to root-cause yet.

It appears non-deterministic.

File: templates/rag-chroma-multi-modal/rag_chroma_multi_modal/ingest.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    n_pages = len(pdf)
+    for page_number in range(n_pages):
+        page = pdf.get_page(page_number)
+        bitmap = page.render(scale=1, rotation=0, crop=(0, 0, 0, 0))
+        pil_image = bitmap.to_pil()
+        pil_image.save(f"{img_dump_path}/img_{page_number + 1}.jpg", format="JPEG")
+
+
+doc_path = Path(__file__).parent / "docs/DDOG_Q3_earnings_deck.pdf"
+img_dump_path = Path(__file__).parent / "docs/"
```
Comment (efriis) [inline templates/rag-chroma-multi-modal/rag_chroma_multi_modal/ingest.py:26]: would it be good to dump images in an inner directory that we can gitignore? Or a different directory?

**Comment by rlancemartin** [comment]
May want to wait until [image prompt template](https://github.com/langchain-ai/langchain/pull/13227) is in.

**Pull request merged**